src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Frame.cs

```diff
@@ -548,6 +551,7 @@ public Task WriteAsync(ArraySegment<byte> data, CancellationToken cancellationTo
             }
 
             _responseBytesWritten += data.Count;
+            VerifyWrite();
```

Comment:
this calling pattern is misleading, _responseBytesWritten gets incremented regardless of if you've written the bytes out. Invert it:
```
             VerifyWrite(data.Count);
             _responseBytesWritten += data.Count;
```

and/or move `_responseBytesWritten += data.Count;` inside of VerifyWrite and make it VerifyAndUpdateWrite